component/encryptor/encryptor.go

```diff
@@ -0,0 +1,368 @@
+	}
+	return path
+}
+
+func (e *Encryptor) StreamDir(options internal.StreamDirOptions) ([]*internal.ObjAttr, string, error) {
+	var objAttrs []*internal.ObjAttr
+	path := formatListDirName(options.Name)
+	log.Info("Encryptor::StreamDir : %s", path)
+	files, err := os.ReadDir(e.mountPointCipher + path)
+	if err != nil {
+		log.Trace("Encryptor::StreamDir : Error reading directory %s : %s", path, err.Error())
+		return nil, "", err
+	}
+
+	for _, file := range files {
+		attr, err := e.GetAttr(internal.GetAttrOptions{Name: path + file.Name()})
+		if err != nil {
+			if err != syscall.ENOENT {
+				log.Trace("Encryptor::StreamDir : Error getting file attributes: %s", err.Error())
+				return objAttrs, "", err
+			}
+			log.Trace("Encryptor::StreamDir : File not found: %s", file.Name())
+			continue
+		}
+
+		objAttrs = append(objAttrs, attr)
+	}
+
+	return objAttrs, "", nil
+}
+
+func checkForActualFileSize(fileHandle *os.File, currentFileSize int64, blockSize int64) (int64, error) {
+
+	log.Info("Encryptor::checkForActualFileSize : currentFileSize %d", currentFileSize)
+	totalBlocks := currentFileSize / (blockSize + MetaSize)
+	if currentFileSize < totalBlocks*(blockSize+MetaSize)+8 {
+		return 0, nil
+	}
+
+	// Read the last 8 bytes of file and check for padding length.
+	paddingLengthBytes := make([]byte, 8)
+
+	// TODO(abhinavgupta) : Find a way to block this read until the last write is done.
+	// While writing to a file if there is a list operation coming in, it will read the
+	// wrong padding length.
+	_, err := fileHandle.ReadAt(paddingLengthBytes, currentFileSize-8)
+	if err != nil {
+		log.Err("Encryptor: Error reading last 8 bytes of file: %s", err.Error())
+		return 0, err
+	}
+
+	actualFileSize := currentFileSize - int64(binary.BigEndian.Uint64(paddingLengthBytes)) - totalBlocks*MetaSize - 8
+	return actualFileSize, nil
+}
+
+func (e *Encryptor) GetAttr(options internal.GetAttrOptions) (attr *internal.ObjAttr, err error) {
+	// Open the file from the mount point and get the attributes.
+	log.Info("Encryptor::GetAttr for %s", options.Name)
+	fileAttr, err := os.Stat(e.mountPointCipher + options.Name)
+	if err != nil {
+		if os.IsNotExist(err) {
+			log.Trace("Encryptor::GetAttr : File not found: %s", options.Name)
+			return nil, syscall.ENOENT
+		}
+		log.Trace("Encryptor::GetAttr : Error getting file attributes: %s", err.Error())
+		return &internal.ObjAttr{}, err
+	}
+
+	// Populate the ObjAttr struct with the file info.
+	attr = &internal.ObjAttr{
+		Mtime:  fileAttr.ModTime(),                // Modified time
+		Atime:  time.Now(),                        // Access time (current time as approximation)
+		Ctime:  fileAttr.ModTime(),                // Change time (same as modified time in this case)
+		Crtime: fileAttr.ModTime(),                // Creation time (not available in Go, using modified time)
+		Mode:   fileAttr.Mode(),                   // Permissions
+		Path:   e.mountPointCipher + options.Name, // Full path
```

Comment:
>// Full path [](http://example.com/codeflow?start=45&length=12)

Should this be just options.Name. The fact that it is in mountPointCipher is an implementation detail. As far as the other components are concerned, they are writing to / reading from options.Name, correct ?